### PR TITLE
#41 Implement PointType Codenjoy

### PR DESCRIPTION
**Pull request by petrokramar**
Add Codenjoy points registration

### CODE REVIEW

File: src/main/java/juja/microservices/gamification/service/AchievementService.java [hunk 1]
```diff
@@ -70,4 +73,40 @@ public String addDaily(String userFromId, String description) {
...
+            throw new UnsupportedAchievementException("Second user cannot be empty");
+        } else if (!"".equals(secondUserToId) && firstUserToId.equalsIgnoreCase(secondUserToId)) {
+            throw new UnsupportedAchievementException("First and second users must be different");
+        } else if (!"".equals(thirdUserToId) && firstUserToId.equalsIgnoreCase(thirdUserToId)) {
+            throw new UnsupportedAchievementException("First and third users must be different");
+        } else if (!"".equals(secondUserToId) && !"".equals(thirdUserToId)
+                && secondUserToId.equalsIgnoreCase(thirdUserToId)) {
+            throw new UnsupportedAchievementException("Second and third users must be different");
+        }
+        Achievement firstThanks = new Achievement(userFromId, firstUserToId, CODENJOY_FIRST_PLACE,
```
Comment (OlgaKulikova) [inline src/main/java/juja/microservices/gamification/service/AchievementService.java]: I think the correct name of achievement is firstPlace, secondPlace, thirdPlace. But not thanks.

File: src/main/java/juja/microservices/gamification/service/AchievementService.java [hunk 2]
```diff
@@ -70,4 +73,40 @@ public String addDaily(String userFromId, String description) {
...
         return result;
     }
+
+    public List<String> addCodenjoy(String  userFromId, String firstUserToId, String secondUserToId,
+            String thirdUserToId) {
+        List<String> result = new ArrayList<>();
+        List<Achievement> codenjoyUsersToday = achievementRepository.getAllCodenjoyAchievementsCurrentDate();
+        if (!codenjoyUsersToday.isEmpty()) {
+            throw new UnsupportedAchievementException("You cannot give codenjoy points twice a day");
+        } else if ("".equals(userFromId)) {
```
Comment (OlgaKulikova) [inline src/main/java/juja/microservices/gamification/service/AchievementService.java:87]: I think the logic about required fuilds should be checked on the slack-bot level, before the request went to the server side. We can discuss it on meeting.

File: src/main/java/juja/microservices/gamification/service/AchievementService.java [hunk 3]
```diff
@@ -70,4 +73,40 @@ public String addDaily(String userFromId, String description) {
...
+        } else if (!"".equals(thirdUserToId) && firstUserToId.equalsIgnoreCase(thirdUserToId)) {
+            throw new UnsupportedAchievementException("First and third users must be different");
+        } else if (!"".equals(secondUserToId) && !"".equals(thirdUserToId)
+                && secondUserToId.equalsIgnoreCase(thirdUserToId)) {
+            throw new UnsupportedAchievementException("Second and third users must be different");
+        }
+        Achievement firstThanks = new Achievement(userFromId, firstUserToId, CODENJOY_FIRST_PLACE,
+                "Codenjoy first place", AchievementType.CODENJOY);
+        result.add(achievementRepository.addAchievement(firstThanks));
+        if (!"".equals(secondUserToId)) {
```
Comment (OlgaKulikova) [inline src/main/java/juja/microservices/gamification/service/AchievementService.java]: We should discuss if it possible to send empty fields to server. (The same question is above)

File: src/main/java/juja/microservices/gamification/controller/AchievementController.java [hunk 4]
```diff
@@ -34,4 +35,12 @@
         List<String> ids = achievementService.addThanks(request.getFrom(), request.getTo(), request.getDescription());
         return ResponseEntity.ok(ids);
     }
+
+    @RequestMapping(value = "/achieve/codenjoy", method = RequestMethod.POST)
+    @ResponseBody
+    public ResponseEntity<?> addCodenjoy(@RequestBody CodenjoyRequest request) {
+        List<String> ids = achievementService.addCodenjoy(request.getFrom(), request.getFirstPlace(),
```
Comment (OlgaKulikova) [inline src/main/java/juja/microservices/gamification/controller/AchievementController.java]: I think you can pass to service the whole object CodenjoyRequest and get fuilds from this object there - on the service layer.

File: src/main/java/juja/microservices/gamification/service/AchievementService.java (hunk 1 above)
Comment (petrokramar) [reply src/main/java/juja/microservices/gamification/service/AchievementService.java]: Done

File: src/main/java/juja/microservices/gamification/controller/AchievementController.java (hunk 4 above)
Comment (petrokramar) [reply src/main/java/juja/microservices/gamification/controller/AchievementController.java]: Done. The same situation with addDaily and addThanks.

**Review by OlgaKulikova — APPROVED** [review]

**Pull request merged**